File: 프로그래머스/3/49191. 순위/순위.py
from collections import defaultdict

def solution(n, results):
    dic_win = defaultdict(set)
    dic_lose = defaultdict(set)
    for _ in range(n):
        for win, lose in results:
            dic_win[win].add(lose)
            dic_win[win] = dic_win[win].union(dic_win[lose])
            dic_lose[lose].add(win)
            dic_lose[lose] = dic_lose[lose].union(dic_lose[win])
    
    ans = 0
    for i in range(1,n+1):
        if len(dic_win[i]) + len(dic_lose[i]) == n-1:
            ans += 1
    
    return ans

File: 프로그래머스/3/49191. 순위/순위.java
import java.util.*;

class Solution {
    
    public int solution(int n, int[][] results) {
        int[][] field = new int[n+1][n+1];
        for(int i = 0; i < results.length; i++){
            field[results[i][0]][results[i][1]] = 1;
            field[results[i][1]][results[i][0]] = -1;
        }
        for(int k = 1; k <= n; k++){
            for(int i = 1; i <= n; i++){
                for(int j = 1; j <= n; j++){
                   if(k!=i && k!=j && j!=i){
                       if(field[i][k] == 1 && field[k][j] == 1 && field[i][j] == 0){
                           field[i][j] = 1;
                       }
                       else if(field[i][k] == -1 && field[k][j] == -1 && field[i][j] == 0){
                           field[i][j] = -1;
                       }
                   }
                }
            }
        }
        int ans = 0;
        for(int i = 1; i <= n; i++){
            boolean tmp = false;
            for(int j = 1; j <= n; j++){
                if(i != j && (field[i][j] == 0 || field[j][i] == 0)){
                    tmp = true;
                    break;
                }
            }
            if(!tmp){
                ans ++;
            }
        }
        
    return ans;
    }
}